File: internal/domain/download/file.go
package download

// FileID identifies a File entity within a Job.
type FileID int64

// File is one logical file split across one or more segments. Files
// belong to a Job and inherit its lifecycle.
type File struct {
	id           FileID
	jobID        JobID
	filename     string
	poster       string
	groups       []string
	sizeBytes    int64
	state        FileState
	segmentCount int
	segmentsDone int
	isPar2       bool
	segments     []*Segment
}

// NewFileParams constructs a fresh pending File along with its
// segments.
type NewFileParams struct {
	Filename  string
	Poster    string
	Groups    []string
	SizeBytes int64
	IsPar2    bool
	Segments  []NewSegmentParams
}

func newFile(p NewFileParams) *File {
	f := &File{
		filename:     p.Filename,
		poster:       p.Poster,
		groups:       append([]string(nil), p.Groups...),
		sizeBytes:    p.SizeBytes,
		state:        FileStatePending,
		segmentCount: len(p.Segments),
		isPar2:       p.IsPar2,
	}
	for _, sp := range p.Segments {
		f.segments = append(f.segments, newSegment(sp))
	}
	return f
}

// HydrateFileParams is what the repository hands back when loading.
type HydrateFileParams struct {
	ID           FileID
	JobID        JobID
	Filename     string
	Poster       string
	Groups       []string
	SizeBytes    int64
	State        FileState
	SegmentCount int
	SegmentsDone int
	IsPar2       bool
	Segments     []*Segment
}

// HydrateFile is the constructor adapters use to reconstruct a File
// from persistence. No events are emitted.
func HydrateFile(p HydrateFileParams) *File {
	return &File{
		id:           p.ID,
		jobID:        p.JobID,
		filename:     p.Filename,
		poster:       p.Poster,
		groups:       append([]string(nil), p.Groups...),
		sizeBytes:    p.SizeBytes,
		state:        p.State,
		segmentCount: p.SegmentCount,
		segmentsDone: p.SegmentsDone,
		isPar2:       p.IsPar2,
		segments:     p.Segments,
	}
}

// Accessors.
func (f *File) ID() FileID         { return f.id }
func (f *File) JobID() JobID       { return f.jobID }
func (f *File) Filename() string   { return f.filename }
func (f *File) Poster() string     { return f.poster }
func (f *File) Groups() []string   { return append([]string(nil), f.groups...) }
func (f *File) SizeBytes() int64   { return f.sizeBytes }
func (f *File) State() FileState   { return f.state }
func (f *File) SegmentCount() int  { return f.segmentCount }
func (f *File) SegmentsDone() int  { return f.segmentsDone }
func (f *File) IsPar2() bool       { return f.isPar2 }
func (f *File) Segments() []*Segment {
	out := make([]*Segment, len(f.segments))
	copy(out, f.segments)
	return out
}

// SetID is called by the repository after insert.
func (f *File) SetID(id FileID) {
	f.id = id
	for _, s := range f.segments {
		s.SetFileID(id)
	}
}

// SetJobID is called when the file is associated with a (newly-saved)
// job row.
func (f *File) SetJobID(id JobID) { f.jobID = id }

File: src/domain/download/file.zig
//! `File` — one logical file of a release, split across one or more
//! segments. Files belong to a `Job` and inherit its lifecycle.
//!
//! Like `Segment`, a File carries no allocator: its `filename`,
//! `poster`, `groups` and `segments` are all allocated and freed by the
//! owning Job.

const std = @import("std");
const state_mod = @import("state.zig");
const segment_mod = @import("segment.zig");

const FileId = state_mod.FileId;
const JobId = state_mod.JobId;
const FileState = state_mod.FileState;
const Segment = segment_mod.Segment;
const NewSegmentParams = segment_mod.NewSegmentParams;

/// Constructor input for a fresh pending File plus its segments.
/// Strings and slices are borrowed for the duration of the call; the
/// Job copies what it keeps.
pub const NewFileParams = struct {
    filename: []const u8,
    poster: []const u8 = "",
    groups: []const []const u8 = &.{},
    size_bytes: i64 = 0,
    is_par2: bool = false,
    is_recovery_vol: bool = false,
    segments: []const NewSegmentParams,
};

/// Constructor input for a File being rebuilt from persistence.
pub const HydrateFileParams = struct {
    id: FileId = 0,
    job_id: JobId = 0,
    filename: []const u8,
    poster: []const u8 = "",
    groups: []const []const u8 = &.{},
    size_bytes: i64 = 0,
    state: FileState = .pending,
    /// Kept as a stored field rather than `segments.len` because the
    /// "shallow" repository queries hydrate a File's metadata without
    /// its segments, and the UI still wants the count.
    segment_count: i32 = 0,
    segments_done: i32 = 0,
    is_par2: bool = false,
    is_recovery_vol: bool = false,
    segments: []const segment_mod.HydrateSegmentParams = &.{},
};

pub const File = struct {
    /// Assigned by the persistence layer on INSERT; 0 until then.
    id: FileId = 0,
    job_id: JobId = 0,
    /// Owned by the Job; immutable once constructed, so `FileCompleted`
    /// borrows it rather than copying.
    filename: []const u8,
    /// Owned by the Job.
    poster: []const u8 = "",
    /// Owned by the Job — both the outer slice and each string.
    groups: []const []const u8 = &.{},
    size_bytes: i64 = 0,
    /// Read-only from outside the aggregate.
    state: FileState = .pending,
    segment_count: i32 = 0,
    segments_done: i32 = 0,
    is_par2: bool = false,
    /// True when the filename matched `<base>.vol###+##.par2`. While a
    /// Job has `fetch_recovery_vols = false` the orchestrator skips
    /// these files' segments entirely — see `Job.pendingSegments`. The
    /// small index PAR2 (no `.vol`) stays false here even though
    /// `is_par2` is true.
    is_recovery_vol: bool = false,
    /// Owned by the Job. Pointers into this slice are stable for the
    /// aggregate's lifetime: it is allocated once at construction and
    /// never resized, which is what lets the segment index cache
    /// `*Segment`.
    segments: []Segment = &.{},

    /// Called by the repository after INSERT. Pushes the new id down
    /// into the segments, as the Go version did.
    pub fn setId(self: *File, id: FileId) void {
        self.id = id;
        for (self.segments) |*s| s.setFileId(id);
    }

    /// Called when the file is associated with a newly-saved job row.
    pub fn setJobId(self: *File, id: JobId) void {
        self.job_id = id;
    }
};
